File: request.go
package main

import (
	// "bytes"
	"fmt"
	// "io"
	"net/http"
	"sync"
	"time"
)

const NUM_REQUESTS int = 100
var bad = false
var asyncBad = false

func main() {
	var wg sync.WaitGroup
	startTime := time.Now().UnixMilli()
	wg.Add(NUM_REQUESTS)
	for range [NUM_REQUESTS]int{} {
		go func() {
			defer wg.Done()
			// req, _ := http.NewRequest(http.MethodPost, "https://www.example.com", bytes.NewReader([]byte("request body")))
			req, _ := http.NewRequest(http.MethodGet, "https://www.example.com", nil)
			client := http.Client{
				Timeout: 1 * time.Second,
			}
			// req.Header.Set("header-name", "value")
			resp, err := client.Do(req)
			if err != nil {
				fmt.Println(err)
				return
			}
			// body, err := io.ReadAll(resp.Body)
			// if err != nil {
			// 	fmt.Println(err)
			// 	return
			// }
			// fmt.Println(string(body))
			// contentType := resp.Header.Get("Content-Type")
			// fmt.Println(contentType)
			if resp.StatusCode != 200 {
				asyncBad = true
			}
		}()
	}
	wg.Wait()
	endTime := time.Now().UnixMilli()
	elapsedTime := float64(endTime - startTime) / 1000.0
	if bad {
		fmt.Println("Had failed asynchronous request")
	} else {
		fmt.Printf("Asynchronous elapsed time for %v requests = %v seconds\n", NUM_REQUESTS, elapsedTime)
	}

	startTime = time.Now().UnixMilli()
	for range [NUM_REQUESTS]int{} {
		// req, _ := http.NewRequest(http.MethodPost, "https://www.example.com", bytes.NewReader([]byte("request body")))
		req, _ := http.NewRequest(http.MethodGet, "https://www.example.com", nil)
		client := http.Client{
			Timeout: 1 * time.Second,
		}
		// req.Header.Set("header-name", "value")
		resp, err := client.Do(req)
		if err != nil {
			fmt.Println(err)
			continue
		}
		// body, err := io.ReadAll(resp.Body)
		// if err != nil {
		// 	fmt.Println(err)
		// 	return
		// }
		// fmt.Println(string(body))
		// contentType := resp.Header.Get("Content-Type")
		// fmt.Println(contentType)
		if resp.StatusCode != 200 {
			bad = true
		}
	}
	endTime = time.Now().UnixMilli()
	elapsedTime = float64(endTime - startTime) / 1000.0
	if bad {
		fmt.Println("Had failed synchronous request")
	} else {
		fmt.Printf(" Synchronous elapsed time for %v requests = %v seconds\n", NUM_REQUESTS, elapsedTime)
	}
}

File: Request.cs
const int NUM_REQUESTS = 100;
bool bad = false;
bool asyncBad = false;

HttpClient client = new()
{
    Timeout = TimeSpan.FromSeconds(1)
};

long startTime = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
var tasks = new List<Task>(NUM_REQUESTS);
for (int i = 0; i < NUM_REQUESTS; i++)
{
    // HttpRequestMessage request = new(HttpMethod.Post, "https://www.example.com")
    // {
    //     Content = new StringContent("request body")
    // };
    // request.Headers.Add("header-name", "value");
    //
    // tasks.Add(Task.Run(() => client.SendAsync(request).ContinueWith(responseTask =>
    //
    // or without a header:
    //
    // tasks.Add(Task.Run(() => client.PostAsync("http://www.example.com/",
    //     new StringContent("request body")).ContinueWith(responseTask =>
    tasks.Add(Task.Run(() => client.GetAsync("http://www.example.com/").ContinueWith(responseTask =>
    {
        HttpResponseMessage response = responseTask.Result;
        if (((int)response.StatusCode) != 200)
        {
            asyncBad = true;
        }
        else
        {
            // string? contentType = null;
            // if (response.Headers.TryGetValues("Content-Type", out IEnumerable<string>? values))
            // {
            //     contentType = values.First();
            // }
            // Console.WriteLine(contentType);
            // string body = response.Content.ReadAsStringAsync().Result;
            // Console.WriteLine(body);
        }
    })));
}
Task.WaitAll(tasks);
long endTime = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
double elapsedTime = (endTime - startTime) / 1000.0;
if (asyncBad)
{
    Console.WriteLine("Had failed asynchronous request");
}
else
{
    Console.WriteLine("Asynchronous elapsed time for " + NUM_REQUESTS + " requests = " + elapsedTime + " seconds");
}

startTime = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
for (int i = 0; i < NUM_REQUESTS; i++)
{
    HttpRequestMessage request = new(HttpMethod.Get, "https://www.example.com");
    // HttpRequestMessage request = new(HttpMethod.Post, "https://www.example.com")
    // {
    //     Content = new StringContent("request body")
    // };
    // request.Headers.Add("header-name", "value");
    HttpResponseMessage response = client.Send(request);
    if (!response.IsSuccessStatusCode) // code is not in range 200-299
    {
        bad = true;
    }
    else
    {
        // string? contentType = response.Content.Headers.ContentType?.ToString();
        // Console.WriteLine(contentType);
        // string body = response.Content.ReadAsStringAsync().Result;
        // Console.WriteLine(body);
    }
}
endTime = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
elapsedTime = (endTime - startTime) / 1000.0;
if (bad)
{
    Console.WriteLine("Had failed synchronous request");
}
else
{
    Console.WriteLine(" Synchronous elapsed time for " + NUM_REQUESTS + " requests = " + elapsedTime + " seconds");
}
